test/test_torch.py

```diff
@@ -2905,6 +2905,56 @@ def test_use_deterministic_algorithms_beta_warning(self, device):
             with self.assertWarnsOnceRegex(UserWarning, "torch.use_deterministic_algorithms is in beta"):
                 torch.use_deterministic_algorithms(True)
 
+    # This test checks that each nondeterministic function throws an error as
+    # described in the documentation for torch.use_deterministic_algorithms
+    def test_use_deterministic_algorithms_errors(self, device):
+
+        # Add new test cases to this list any time a nondeterministic error is
+        # added to an operation.
+        #
+        # name                  Name of test case
+        #
+        # test_func             Forward pass function that exercises the
+        #                       nondeterministic operation
+        #
+        # error_device          Device that should throw the error.
+        #                       Set to None if all devices should throw an error.
+        #
+        # caller_name_forward   Name of function shown in error for forward pass.
+        #                       Set to None if the forward should not throw an error.
+        #
+        # caller_name_backward  Name of function shown in error for backward pass.
+        #                       Set to None if the backward should not throw an error.
+        nondeterministic_operations = [
+            # name, test_func, error_device, caller_name_forward, caller_name_backward
+            (
+                'kthvalue',
+                (lambda slf, device: torch.randn(10, device=device).kthvalue(4)),
+                'cuda', 'kthvalue CUDA', None
+            ),
+            (
+                'index_select_backward',
+                (lambda slf, device: torch.randn(10, device=device, requires_grad=True).index_select(0, torch.tensor([3], device=device))),
+                'cuda', None, 'index_add_cuda_'
+            ),
+            (
+                'index_copy',
+                (lambda slf, device: torch.randn(10, device=device).index_copy(0, torch.tensor([3], device=device), torch.randn(1, device=device))),
```

Comment:
This might not be a very good design. If the test function needs to be multiple lines, then it will need to be defined up above. Eventually, we might have many test functions all packed into `test_use_deterministic_algorithms_errors`. That's not necessarily wrong, but it could get difficult to keep track of indentations.

Reply:
This is a cool design for a tricky problem.

Would it just be simpler to create two tests: test_use_deterministic_algorithms_cpu_errors and test_use_deterministic_algorithms_cuda_errors and just manually create the input to trigger each determinism warning? The first test would have the @onlyCPU decorator and the second would have the @onlyCUDA decorator.

This alone would still be a little verbose because each op would need a line to create the input, a line to use the context manager to catch error, and a line to call the op. But that seems OK? Would that really be many more LoCs than trying a more structured approach like this one? 

Also happy to talk about ideas more on Slack.

Reply:
Yeah I guess it's not too difficult to manage as it is.

I'm not sure I understand the benefit of splitting cpu and cuda into two different tests. Some errors happen for both devices, so I figure it's best to keep them combined

Reply:
That's a good point. Sections of the test can also guard on the device type being CPU or CUDA if that's easier.